### PR TITLE
Add streaming support for PKCS7_VerifySignedData.

### PR DESCRIPTION
**Pull request by TakayukiMatsuo**
# Description

Extend PKCS7_VerifySignedData streaming to support  PKCS7 bundle which has multiple part octet.

Fixes zd#16606

# Testing

Run an application and data provided in zd#16606.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by TakayukiMatsuo** [comment]
Hi　@JacobBarthelmeh ,

In this PR, I made a small change to wc_PKCS7_AddDataToStream. This change resets pksc7->stream>idx to 0 after reading the input data that satisfies the argument expected. Without this modification, If pksc7->stream>idx is not reset to 0, the input data will be skipped in the next wc_PKCS7_AddDataToStream call.

I thought this was a bug, so made a change in this PR. However, this change caused errors in existing streaming test cases.
Am wondering if this change is wrong, or if I'm calling wc_PKCS7_AddDataToStream incorrectly. I would be happy to hear your opinions on the changes I have made.

Thanks,
Tak

File: wolfcrypt/src/pkcs7.c [hunk 1]
```diff
@@ -4890,198 +5282,178 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
-                }
-                else {
-                    XMEMCPY(pkcs7->stream->content, content, contentSz);
-                    pkcs7->stream->contentSz = contentSz;
-                }
+            if (pkcs7->stream->length > 0) {
+                localIdx = 0;
             }
+            multiPart = pkcs7->stream->multi;
+            detached  = pkcs7->stream->detached;
```
Comment (douzzer) [inline wolfcrypt/src/pkcs7.c:5402]: These assignments are producing
```
wolfssl/wolfcrypt/src/pkcs7.c:5403:13: warning: Value stored to 'multiPart' is never read [clang-analyzer-deadcode.DeadStores]
wolfssl/wolfcrypt/src/pkcs7.c:5404:13: warning: Value stored to 'detached' is never read [clang-analyzer-deadcode.DeadStores]
```

**Review by douzzer — CHANGES_REQUESTED** [review]
Hi Tak.  Most of the change requests from my review are for reports from automated analysis.

The one bigger problem is regarding the change to `wc_PKCS7_AddDataToStream()`.  In order to get the existing unit tests to pass, I had to comment that out.

I don't yet understand the problem you were addressing with that clause, but whatever it is, we need to include new test(s) in `wolfcrypt/test/test.c` and/or `tests/api.c` that test for it and fail without the fix.

**Comment by JacobBarthelmeh** [comment]
FYI for the current failing PRB master it looks to be a build when PKCS7 streaming mode is disabled (defining the macro NO_PKCS7_STREAM):

```
wolfcrypt/src/pkcs7.c:5273:18: error: ‘PKCS7’ has no member named ‘stream’
 5273 |         if (pkcs7->stream->content != NULL) {

      |                  ^~

In file included from ./wolfssl/wolfcrypt/pkcs7.h:29,

                 from wolfcrypt/src/pkcs7.c:31:

wolfcrypt/src/pkcs7.c:5274:24: error: ‘PKCS7’ has no member named ‘stream’
 5274 |             XFREE(pkcs7->stream->content, pkcs7->heap, DYNAMIC_TYPE_PKCS7);
      |                        ^~
./wolfssl/wolfcrypt/types.h:573:63: note: in definition of macro ‘XFREE’
```

Not as sure on the krb5 actions test, it would need farther investigation.

**Comment by TakayukiMatsuo** [comment]
@douzzer , @JacobBarthelmeh  and @dgarske ,  thank you for your help. It's been a while, but I will adjust my time to complete this PR.

File: wolfcrypt/src/pkcs7.c [hunk 2]
```diff
@@ -4890,198 +5282,178 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
+
+                /* if certificates set has indefinite length, try to get
+                 * the first certificate length of the set.
+                 */
+                if (ret == 0 && pkiMsg2[localIdx] == ASN_INDEF_LENGTH) {
+
+                    localIdx++;
+                    certIdx = localIdx;
+                #ifndef NO_PKCS7_STREAM
+                    /* set indef-length count. used for skipping trailing zeros */
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: fixed the length.

File: wolfcrypt/src/pkcs7.c [hunk 3]
```diff
@@ -5108,20 +5480,37 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
+                src     = in;
+                srcSz   = inSz;
+                pkiMsg2 = in;
+            }
+
+            if ((ret = wc_PKCS7_AddDataToStream(pkcs7, src, srcSz,
+                            pkcs7->stream->expected, &pkiMsg2, &idx)) != 0) {
                 break;
             }
+            pkiMsg2Sz = (pkcs7->stream->length > 0)? pkcs7->stream->length:srcSz;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: fixed the length.

File: wolfcrypt/src/pkcs7.c [hunk 4]
```diff
@@ -5329,24 +5696,38 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
-            if (pkcs7->stream->flagOne) {
-                pkiMsg2 = pkiMsg;
 
-                /* check if using internal stream buffer and should adjust sz */
-                if (pkiMsg != in && pkcs7->stream->length > 0) {
-                    pkiMsg2Sz = pkcs7->stream->length;
-                }
-            }
+            /* check if using internal stream buffer and should adjust sz */
+            pkiMsg2Sz = (pkcs7->stream->length > 0)? pkcs7->stream->length:srcSz;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: fixed the length.

File: wolfcrypt/src/pkcs7.c [hunk 5]
```diff
@@ -5413,31 +5812,37 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
-                if (pkiMsg != in && pkcs7->stream->length > 0) {
-                    pkiMsg2Sz = pkcs7->stream->length;
-                }
-                else {
-                    /* if pkiMsg2 is pkiMsg and not using an internal stream
-                     * buffer then the size is limited by inSz */
-                    pkiMsg2Sz = inSz;
-                }
-            }
+            pkiMsg2Sz = (pkcs7->stream->length > 0)? pkcs7->stream->length:srcSz;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: fixed the length.

File: wolfcrypt/src/pkcs7.c (hunk 1 above)
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: removed those two lines.

File: wolfcrypt/src/pkcs7.c [hunk 6]
```diff
@@ -288,6 +304,10 @@ static int wc_PKCS7_AddDataToStream(PKCS7* pkcs7, byte* in, word32 inSz,
         pkcs7->stream->length  += len;
         pkcs7->stream->idx     += len;
         pkcs7->stream->totalRd += len;
+
+        if (pkcs7->stream->idx == inSz) {
+            pkcs7->stream->idx = 0;
+        }
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: removed the portion.

**Comment by bandi13** [comment]
retest this please

**Comment by cconlon** [comment]
@TakayukiMatsuo Is this ready for re-review?

Thanks,
Chris

**Comment by TakayukiMatsuo** [comment]
@cconlon , yes, it is ready.

**Comment by cconlon** [comment]
Thanks @TakayukiMatsuo, I'm assigning over to @JacobBarthelmeh for re-review.  @JacobBarthelmeh, feel free to have @douzzer re-review his comments if you would like.

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]
There appears to be some issues when trying this out. Could you let me know if these tests are invalid?

Steps to recreate:

Create a test SMIME bundle
```
cd wolfssl-examples/pkcs7
openssl smime -sign -in ~/Documents/wolfssl/configure.ac -out test-signed -signer ../certs/ca-cert.pem -nodetach -nocerts -binary -outform DER -stream -inkey ../certs/ca-key.pem 
```

Alter the pkcs7-verify.c example to take in an argument and verify in chunks:

```
--- a/pkcs7/pkcs7-verify.c
+++ b/pkcs7/pkcs7-verify.c
@@ -42,7 +42,7 @@ int main(int argc, char** argv)
 #endif
 
     /* load DER PKCS7 */
-    derFile = fopen("signed.p7s", "rb");
+    derFile = fopen(argv[1], "rb");
     if (derFile) {
         fseek(derFile, 0, SEEK_END);
         derSz = (int)ftell(derFile);
@@ -70,9 +70,14 @@ int main(int argc, char** argv)
     if (rc != 0) goto exit;
     rc = wc_PKCS7_InitWithCert(&pkcs7, NULL, 0);
     if (rc != 0) goto exit;
-    rc = wc_PKCS7_VerifySignedData(&pkcs7, derBuf, derSz);
-    if (rc != 0) goto exit;
 
+    int z, chunkSz = 1;
+    for (z = 0; z < derSz;) {
+        int sz = (z + chunkSz > derSz)? derSz - z : chunkSz;
+        rc = wc_PKCS7_VerifySignedData(&pkcs7, derBuf+z, sz);
+        z+=sz;
+    }
+    if (rc != 0) goto exit;
     printf("PKCS7 Verify Success\n");
 
 exit:
```

With chunkSz set to 1 it fails out with -270 (want more data), but set to 10 it succeeds with -272 (needs sig verify) which is expected here since -nocerts was used when creating the test bundle.


Moving on to a bundle that has certificates in it:

```
openssl smime -sign -in ~/Documents/wolfssl/configure.ac -out test-signed -signer ../certs/ca-cert.pem -nodetach -binary -outform DER -stream -inkey ../certs/ca-key.pem 
```

The test application is getting -270 with chunkSz set to 10:

```
bash-3.2$ ./pkcs7-verify test-signed 
Der 318126
RC=-270
```

Upping chunkSz to 100 then saw it verify successfully.

File: wolfcrypt/src/pkcs7.c [hunk 7]
```diff
@@ -4959,58 +5310,83 @@ static int PKCS7_VerifySignedData(PKCS7* pkcs7, const byte* hashBuf,
...
-                            pkcs7->stream->expected, &pkiMsg, &idx)) != 0) {
+            ret = wc_PKCS7_HandleOctetStrings(pkcs7, in, inSz,
+                                                &stateIdx, &idx, keepContent);
+            if (ret != 0)
                 break;
+
+            /* copy content to pkcs7->contentDynamic */
+            if (keepContent && pkcs7->stream->content &&
+                                            pkcs7->stream->contentSz >0) {
+                pkcs7->contentDynamic   = XMALLOC(pkcs7->stream->contentSz,
```
Comment (douzzer) [inline wolfcrypt/src/pkcs7.c:5363]: This is warning on C++:
```
[all-g++] [7 of 31] [2bce5fef9f]
    configure...   real 0m21.405s  user 0m8.403s  sys 0m14.967s
    build...In file included from ./wolfssl/wolfcrypt/pkcs7.h:29,
                 from wolfcrypt/src/pkcs7.c:31:
wolfcrypt/src/pkcs7.c: In function ‘int PKCS7_VerifySignedData(PKCS7*, const byte*, word32, byte*, word32, byte*, word32)’:
9386a882b9 (<juliusz@wolfssl.com> 2021-10-19 15:51:29 +0200 571)                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
./wolfssl/wolfcrypt/types.h:571:67: error: invalid conversion from ‘void*’ to ‘byte*’ {aka ‘unsigned char*’} [-fpermissive]
  571 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*
2bce5fef9f (<tak@wolfssl.com> 2023-11-07 13:09:38 +0900 5363)                 pkcs7->contentDynamic   = XMALLOC(pkcs7->stream->contentSz,
wolfcrypt/src/pkcs7.c:5363:43: note: in expansion of macro ‘XMALLOC’
 5363 |                 pkcs7->contentDynamic   = XMALLOC(pkcs7->stream->contentSz,
      |                                           ^~~~~~~
make[2]: *** [Makefile:7251: wolfcrypt/src/src_libwolfssl_la-pkcs7.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:8064: all-recursive] Error 1
make: *** [Makefile:4782: all] Error 2
   real 0m14.125s  user 1m20.695s  sys 0m8.076s
    all-g++ fail_build
    failed config: '--enable-all' '--enable-testcert' '--enable-srtp' '--enable-sp-math-all' 'CC=g++' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK'
```

**Review by douzzer — CHANGES_REQUESTED** [review]
looks mostly good -- `super-quick-check` found one trivial defect.

File: wolfcrypt/src/pkcs7.c (hunk 7 above)
Comment (TakayukiMatsuo) [reply wolfcrypt/src/pkcs7.c]: Added a cast to XMALLOC.

**Comment by TakayukiMatsuo** [comment]
Hi Jacob,  

Probably need to modify the code fragment you use a bit to be able to handle the return code from wc_PKCS7_VerifySignedData, like this:
```
 for (z = 0; z < derSz;) {
        int sz = (z + chunkSz > derSz)? derSz - z : chunkSz;
        rc = wc_PKCS7_VerifySignedData(&pkcs7, derBuf + z, sz);
        if (rc < 0) {
            if (rc == WC_PKCS7_WANT_READ_E) {
                z += sz;
                continue;
            }
            printf("Error z:%d %d: %s\n", z, rc, wc_GetErrorString(rc));
            break;
        }
        else {
            break;
        }
    }
```

With this code, could get expected result with the test bundle file.
Tak

**Comment by JacobBarthelmeh** [comment]
> Hi Jacob,
> 
> Probably need to modify the code fragment you use a bit to be able to handle the return code from wc_PKCS7_VerifySignedData, like this:
> 
> ```
>  for (z = 0; z < derSz;) {
>         int sz = (z + chunkSz > derSz)? derSz - z : chunkSz;
>         rc = wc_PKCS7_VerifySignedData(&pkcs7, derBuf + z, sz);
>         if (rc < 0) {
>             if (rc == WC_PKCS7_WANT_READ_E) {
>                 z += sz;
>                 continue;
>             }
>             printf("Error z:%d %d: %s\n", z, rc, wc_GetErrorString(rc));
>             break;
>         }
>         else {
>             break;
>         }
>     }
> ```
> 
> With this code, could get expected result with the test bundle file. Tak

Tak, adding an early bail on error case does succeed. This likely means that the last bytes of the bundle are not being parsed though and that it is reaching the (-272) error in this case early. With indef it should only be 0's but I think we should have it confirm that they are. Adding `derBuf[derSz - 2] = derBuf[derSz - 2] + 1;  ` before verifying the bundle seems to still return success.

**Comment by bandi13** [comment]
retest this please

**Comment by TakayukiMatsuo** [comment]
Hi Jacob, added trailing zero's check.

**Comment by JacobBarthelmeh** [comment]
Hi @TakayukiMatsuo , Please double check the case of altering the last bytes in the bundle before verification.

**Comment by TakayukiMatsuo** [comment]
Hi @JacobBarthelmeh , Confirmed ASN_PARSE_E(-140) is returned when the last byte of the bundle is altered by adding  `derBuf[derSz - 2] = derBuf[derSz - 2] + 1; `  to the sample code.

**Review by JacobBarthelmeh — APPROVED** [review]
Thanks @TakayukiMatsuo for all your hard work on this!

**Pull request merged**